stable/database/files/nuosm

```diff
@@ -295,27 +293,49 @@ function perform_restore() {
 }
 
 #=======================================
-# function - retry the provided function
-#
-# Retry the passed function with arguments until succees or max retry count is
-# reached.
+# function - retry the supplied command until it succeeds or the supplied
+# timeout (in seconds) expires.
 #
 function retry() {
-  local max="$1"
+  # Calculate deadline from timeout
+  local timeout="$1"
+  local start="$(date -u '+%s')"
+  local deadline="$((start + timeout))"
   shift
-  local count=0
+
+  # Start with 1 second interval and increase exponentially
   local rc=0
-  while [ "$count" -lt "$max" ]; do
-    [ $count -gt 0 ] && log "'$*' exited with code $rc - retrying ..."
-    "$@"
-    rc=$?
-    if [ $rc -eq 0 ]; then
-      return $rc
+  local interval=1
+  local failures=0
+  while true; do
+    # Execute command and check exit code
+    out="$("$@" 2>&1)"
+    rc="$?"
+    if [ "$rc" -eq 0 ]; then
+      return 0
     fi
-    count=$(( count + 1 ))
-    sleep 1
+
+    # Check that timeout did not expire and that we have executed the command
+    # more than once, in case the execution time of the initial command
+    # invocation consumed the entire timeout.
+    failures="$((failures + 1))"
+    if [ "$deadline" -le "$(date -u '+%s')" ] && [ "$failures" -gt 1 ]; then
+      log "'$*' failed with exit code $rc and output [$out] after $timeout seconds and $failures tries"
+      return "$rc"
+    fi
+
+    # Sleep for polling interval and calculate next interval
+    log "'$*' failed with exit code $rc and output [$out] - retrying in $interval seconds..."
+    sleep "$interval"
+    interval="$((interval * 2))"
+
+    # Clamp interval to 10 seconds or seconds until the timeout expires
+    [ "$interval" -lt 10 ] || interval=10
+    local current="$(date -u '+%s')"
+    local remaining="$((deadline - current))"
+    [ "$interval" -lt "$remaining" ] || interval="$remaining"
+    [ "$interval" -ge 1 ] || interval=1
```

Comment:
Wouldn't it make sense to cap the interval to the remaining time just before before sleeping? Otherwise you are doing an execution and then  sleeping for remaining before execution.

Reply:
Makes sense. I updated the polling loop to clamp the value before sleeping.